lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/ldap/JdkLdapClient.java

```diff
@@ -94,13 +99,31 @@ public <T> T processLdapContext(String userName, String password, LdapContextPro
     }
 
     @Override
-    public <T> T executeLdapQuery(String userName, String password, LdapQuery ldapQuery, LdapSearchResultProcessor<T> resultProcessor)
+    public synchronized <T> T executeLdapQuery(String userName, String password, LdapQuery ldapQuery, LdapSearchResultProcessor<T> resultProcessor)
             throws NamingException
     {
-        try (CloseableContext context = createUserDirContext(userName, password);
-                CloseableSearchResults search = searchContext(ldapQuery, context)) {
-            return resultProcessor.process(search.searchResults);
+        final int maxRetry = 2;
+        for (int retry = 0; retry <= maxRetry; retry++) {
+            CloseableContext context = ctxs.get(userName);
+            if (context == null) {
+                context = createUserDirContext(userName, password);
+                ctxs.put(userName, context);
+            }
+            try {
+                try (CloseableSearchResults search = searchContext(ldapQuery, context)) {
+                    return resultProcessor.process(search.searchResults);
+                }
+            }
+            catch (NamingException e) {
+                log.warn("executeLdapQuery failed at attempt {}.", retry, e);
+                if (retry >= maxRetry) {
+                    throw e;
+                }
+            }
+            context.close();
+            ctxs.remove(userName);
```

Comment:
Why we remove the close an remove the context every time?

Reply:
It's for retrying. Let me move it into the catch clause to reduce confusion.